File: src/main/java/com/trading/kalyani/KTManager/service/BrahmastraService.java
package com.trading.kalyani.KTManager.service;

import com.trading.kalyani.KTManager.dto.brahmastra.*;
import com.trading.kalyani.KTManager.entity.BrahmastraSignal;

import java.time.LocalDate;
import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * Service interface for Brahmastra Triple Confirmation Trading Strategy.
 *
 * The strategy combines:
 * 1. Supertrend (ATR-based) - For trend direction
 * 2. MACD (12,26,9) - For momentum confirmation
 * 3. VWAP - For value area confirmation
 * 4. PCR (optional) - For market bias filtering
 */
public interface BrahmastraService {


    // ==================== Signal Generation ====================

    /**
     * Generate trading signals for a given symbol and timeframe.
     * Signals are generated when all three indicators confirm (Supertrend, MACD, VWAP).
     */
    List<SignalDTO> generateSignals(SignalRequest request);

    /**
     * Async version of signal generation for heavy computations.
     */
    CompletableFuture<List<SignalDTO>> generateSignalsAsync(SignalRequest request);

    /**
     * Scan for live signals across multiple symbols.
     */
    List<LiveScanResult> scanLiveSignals(List<String> symbols);

    /**
     * Scan a single symbol for current signal.
     */
    LiveScanResult scanSymbol(String symbol, String timeframe);

    // ==================== Backtesting ====================

    /**
     * Run backtest for a given symbol with specified parameters.
     */
    BacktestResult runBacktest(BacktestRequest request);

    /**
     * Async version of backtest for heavy computations.
     */
    CompletableFuture<BacktestResult> runBacktestAsync(BacktestRequest request);

    // ==================== Dashboard ====================

    /**
     * Get aggregated dashboard summary across all assets.
     */
    DashboardSummary getDashboardSummary();

    /**
     * Get symbol-specific summary.
     */
    SymbolSummary getSymbolSummary(String symbol);

    // ==================== Signal Management ====================

    /**
     * Get active signals.
     */
    List<SignalDTO> getActiveSignals();

    /**
     * Get signals for a date range.
     */
    List<SignalDTO> getSignalsByDateRange(String symbol, LocalDate from, LocalDate to);

    /**
     * Update signal status (for trade management).
     */
    BrahmastraSignal updateSignalStatus(Long signalId, String status, Double exitPrice, String exitReason);

    /**
     * Get signal by ID.
     */
    SignalDTO getSignalById(Long id);

    // ==================== Real-time Scanning ====================

    /**
     * Process live tick for signal detection.
     * Called from DailyJobServiceImpl during job execution.
     */
    void processLiveTick(Long instrumentToken, Double ltp, Integer appJobConfigNum);

    /**
     * Check if triple confirmation is met for current candle.
     */
    boolean isTripleConfirmation(Long instrumentToken, String signalType);

    // ==================== PCR Integration ====================

    /**
     * Get current PCR for a symbol.
     */
    Double getCurrentPCR(String symbol);

    /**
     * Determine market bias from PCR.
     */
    String getPCRBias(Double pcr);

    /**
     * Check if signal is allowed based on PCR bias.
     */
    boolean isSignalAllowedByPCR(String signalType, Double pcr);

    // ==================== Indicator Metrics ====================

    /**
     * Get real-time indicator metrics for a symbol.
     * Returns current values and historical data for Supertrend, MACD, and VWAP.
     */
    IndicatorMetrics getIndicatorMetrics(String symbol, String timeframe, int historyBars);

    /**
     * Get indicator metrics for all tracked symbols.
     */
    List<IndicatorMetrics> getAllIndicatorMetrics(String timeframe, int historyBars);

    // ==================== Option Chain Integration ====================

    /**
     * Get option chain metrics including Max Pain, OI Analysis, and Gamma Exposure.
     * Provides additional confirmation signals for the Brahmastra strategy.
     */
    OptionChainMetrics getOptionChainMetrics(String symbol);

    /**
     * Get option chain metrics for all tracked symbols.
     */
    List<OptionChainMetrics> getAllOptionChainMetrics();

    /**
     * Check if option chain data confirms the trading signal.
     * Uses Max Pain, OI changes, and Gamma Exposure for validation.
     */
    boolean doesOptionChainConfirmSignal(String symbol, String signalType);

    /**
     * Get combined signal from option chain analysis.
     * Returns BUY, SELL, or NEUTRAL based on Max Pain, OI, and GEX.
     */
    String getOptionChainSignal(String symbol);
}

File: kpn-ui/src/app/services/brahmastra.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, Subject } from 'rxjs';
import {
  SignalRequest,
  BacktestRequest,
  SignalResponse,
  BacktestResponse,
  LiveScanResponse,
  DashboardResponse,
  PCRResponse,
  SignalDTO,
  LiveScanResult,
  IndicatorMetricsResponse,
  AllIndicatorMetricsResponse,
  OptionChainMetricsResponse,
  AllOptionChainMetricsResponse,
  OptionChainConfirmationResponse
} from '../models/brahmastra.model';
import { DataService } from './data.service';
import { WebsocketService } from './web-socket.service';

@Injectable({
  providedIn: 'root'
})
export class BrahmastraService {

  // Subject for WebSocket signals
  private signalSubject = new Subject<LiveScanResult>();
  public signalUpdates$ = this.signalSubject.asObservable();

  constructor(
    private dataService: DataService,
    private webSocketService: WebsocketService
  ) {
    // Subscribe to WebSocket Brahmastra signals
    this.webSocketService.getBrahmastraSignals().subscribe({
      next: (signal) => {
        this.signalSubject.next(signal);
      },
      error: (err) => console.error('Error receiving Brahmastra signal:', err)
    });
  }

  // ==================== Signal Generation ====================

  /**
   * Generate trading signals for a given symbol and timeframe.
   */
  generateSignals(request: SignalRequest): Observable<SignalResponse> {
    return this.dataService.generateBrahmastraSignals(request);
  }

  /**
   * Generate signals asynchronously (for large date ranges).
   */
  generateSignalsAsync(request: SignalRequest): Observable<SignalResponse> {
    return this.dataService.generateBrahmastraSignals(request);
  }

  // ==================== Backtesting ====================

  /**
   * Run backtest with specified parameters.
   */
  runBacktest(request: BacktestRequest): Observable<BacktestResponse> {
    return this.dataService.runBrahmastraBacktest(request);
  }

  /**
   * Run backtest asynchronously (for large date ranges).
   */
  runBacktestAsync(request: BacktestRequest): Observable<BacktestResponse> {
    return this.dataService.runBrahmastraBacktest(request);
  }

  // ==================== Live Scanning ====================

  /**
   * Scan multiple symbols for live signals.
   */
  scanLive(symbols: string = 'NIFTY'): Observable<LiveScanResponse> {
    return this.dataService.scanBrahmastraLive(symbols);
  }

  /**
   * Scan a single symbol.
   */
  scanSymbol(symbol: string, timeframe: string = '5m'): Observable<any> {
    return this.dataService.scanBrahmastraSymbol(symbol, timeframe);
  }

  // ==================== Dashboard ====================

  /**
   * Get aggregated dashboard summary.
   */
  getDashboardSummary(): Observable<DashboardResponse> {
    return this.dataService.getBrahmastraDashboard();
  }

  /**
   * Get symbol-specific summary.
   */
  getSymbolSummary(symbol: string): Observable<any> {
    return this.dataService.getBrahmastraSymbolSummary(symbol);
  }

  // ==================== Signal Management ====================

  /**
   * Get all active signals.
   */
  getActiveSignals(): Observable<any> {
    return this.dataService.getActiveBrahmastraSignals();
  }

  /**
   * Get signal history for a symbol.
   */
  getSignalHistory(symbol: string, fromDate: string, toDate: string): Observable<any> {
    return this.dataService.getBrahmastraSignalHistory(symbol, fromDate, toDate);
  }

  /**
   * Get signal by ID.
   */
  getSignalById(id: number): Observable<any> {
    return this.dataService.getBrahmastraSignalById(id);
  }

  /**
   * Update signal status.
   */
  updateSignalStatus(id: number, status: string, exitPrice?: number, exitReason?: string): Observable<any> {
    return this.dataService.updateBrahmastraSignalStatus(id, status, exitPrice, exitReason);
  }

  // ==================== PCR Integration ====================

  /**
   * Get current PCR values for all symbols.
   */
  getCurrentPCR(): Observable<PCRResponse> {
    return this.dataService.getBrahmastraPCR();
  }

  // ==================== Indicator Metrics ====================

  /**
   * Get real-time indicator metrics for a specific symbol.
   */
  getIndicatorMetrics(symbol: string, timeframe: string = '5m', historyBars: number = 50): Observable<IndicatorMetricsResponse> {
    return this.dataService.getBrahmastraIndicatorMetrics(symbol, timeframe, historyBars);
  }

  /**
   * Get indicator metrics for all tracked symbols.
   */
  getAllIndicatorMetrics(timeframe: string = '5m', historyBars: number = 50): Observable<AllIndicatorMetricsResponse> {
    return this.dataService.getAllBrahmastraIndicatorMetrics(timeframe, historyBars);
  }

  // ==================== Option Chain Integration ====================

  /**
   * Get option chain metrics (Max Pain, OI Analysis) for a symbol.
   */
  getOptionChainMetrics(symbol: string): Observable<OptionChainMetricsResponse> {
    return this.dataService.getBrahmastraOptionChainMetrics(symbol);
  }

  /**
   * Get option chain metrics for all tracked symbols.
   */
  getAllOptionChainMetrics(): Observable<AllOptionChainMetricsResponse> {
    return this.dataService.getAllBrahmastraOptionChainMetrics();
  }

  /**
   * Check if option chain data confirms a signal.
   */
  checkOptionChainConfirmation(symbol: string, signalType: string): Observable<OptionChainConfirmationResponse> {
    return this.dataService.checkBrahmastraOptionChainConfirmation(symbol, signalType);
  }

  // ==================== Health Check ====================

  /**
   * Check if Brahmastra service is running.
   */
  healthCheck(): Observable<any> {
    return this.dataService.getBrahmastraHealth();
  }

  // ==================== WebSocket Handling ====================

  /**
   * Handle incoming WebSocket signal.
   */
  handleWebSocketSignal(signal: LiveScanResult): void {
    this.signalSubject.next(signal);
  }

  // ==================== Utility Methods ====================

  /**
   * Format date for API request.
   */
  formatDate(date: Date): string {
    return date.toISOString().split('T')[0];
  }

  /**
   * Get default signal request.
   */
  getDefaultSignalRequest(): SignalRequest {
    const today = new Date();
    const weekAgo = new Date(today);
    weekAgo.setDate(weekAgo.getDate() - 7);

    return {
      symbol: 'NIFTY',
      timeframe: '5m',
      fromDate: this.formatDate(weekAgo),
      toDate: this.formatDate(today),
      usePCR: true,
      supertrendPeriod: 20,
      supertrendMultiplier: 2.0,
      macdFastPeriod: 12,
      macdSlowPeriod: 26,
      macdSignalPeriod: 9,
      vwapTolerance: 0.002
    };
  }

  /**
   * Get default backtest request.
   */
  getDefaultBacktestRequest(): BacktestRequest {
    const today = new Date();
    const monthAgo = new Date(today);
    monthAgo.setDate(monthAgo.getDate() - 30);

    return {
      symbol: 'NIFTY',
      timeframe: '5m',
      fromDate: this.formatDate(monthAgo),
      toDate: this.formatDate(today),
      usePCR: true,
      initialCapital: 100000,
      riskPerTrade: 1,
      supertrendPeriod: 20,
      supertrendMultiplier: 2.0,
      macdFastPeriod: 12,
      macdSlowPeriod: 26,
      macdSignalPeriod: 9,
      vwapTolerance: 0.002,
      useTrailingStop: false,
      usePartialProfits: true,
      partialProfitPercent: 50
    };
  }
}

